### PR TITLE
[Epix] Add new extractor

### LINKED ISSUE #3185: EPIX Support
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.03.08.1**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

_No response_

### Example URLs

Single Video - https://www.epix.com/series/from/season/1/episode/1/from-s1-e1
Single Video - https://www.epix.com/series/from/season/1/episode/7/from-s1-e7?trailer=true
Single Video - https://www.epix.com/series/from/extra/13596

### Description

N/A

### Verbose log

```shell
[debug] Command-line config: ['-vU', 'https://www.epix.com/series/from/season/1/episode/7/from-s1-e7']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] yt-dlp version 2022.03.08.1 [c0c2c57] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] exe versions: ffmpeg 5.0-full_build-www.gyan.dev (setts), ffprobe 5.0-full_build-www.gyan.dev
[debug] Optional libraries: brotli, Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.03.08.1, Current version: 2022.03.08.1
yt-dlp is up to date (2022.03.08.1)
[debug] [generic] Extracting URL: https://www.epix.com/series/from/season/1/episode/7/from-s1-e7
[generic] from-s1-e7: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] from-s1-e7: Downloading webpage
[generic] from-s1-e7: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://www.epix.com/series/from/season/1/episode/7/from-s1-e7
Traceback (most recent call last):
  File "yt_dlp\YoutubeDL.py", line 1389, in wrapper
  File "yt_dlp\YoutubeDL.py", line 1459, in __extract_info
  File "yt_dlp\extractor\common.py", line 617, in extract
  File "yt_dlp\extractor\generic.py", line 3985, in _real_extract
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.epix.com/series/from/season/1/episode/7/from-s1-e7
```

### PR DESCRIPTION
**Pull request by hatienl0i261299**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Close #3185

### CODE REVIEW

File: yt_dlp/extractor/epix.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:3]: remove (this will fix the merge conflict)

File: yt_dlp/extractor/epix.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+class EpixBase(InfoExtractor):
+    _API_URL = 'https://api.epix.com/v2/'
+    _API_KEY = 'f07debfcdf0f442bab197b517a5126ec'
+    _HEADERS = {
+        'user-agent': 'Mozilla/5.0 (X11; Linux i686) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/65.0.3325.181 Chrome/65.0.3325.181 Safari/537.36',
+        'content-type': 'application/json'
+    }
+
+    def _real_initialize(self):
+        info = self._download_json(f'{self._API_URL}sessions', video_id='access token', data=json.dumps({
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:24]:
```suggestion
        info = self._download_json(f'{self._API_URL}sessions', None, data=json.dumps({
```

File: yt_dlp/extractor/epix.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            },
+            'apikey': self._API_KEY,
+            'oauth': {
+                'token': None
+            }
+        }).encode('utf-8'), headers=self._HEADERS, note='Download access token info')
+        access_token = traverse_obj(info, ('device_session', 'session_token'))
+        self._HEADERS.update({
+            'x-session-token': access_token
+        })
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:41]:
```py
        info = self._download_json(
            f'{self._API_URL}sessions', None, note='Download access token info',
            headers=self._HEADERS, note= data=json.dumps({
                ...
            }).encode())
        self._HEADERS.update({
            'x-session-token': info[...]
        })
```

File: yt_dlp/extractor/epix.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+
+class EpixExtraIE(EpixBase):
+    _VALID_URL = r'https?://www\.epix\.com/series/(?P<id>[^(?:/|\?)]+)(?:\/extra\/\d+|\?|$)'
+    IE_NAME = 'epix:extra'
+    IE_DESC = 'epix.com'
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:47]: this is not sufficient description

File: yt_dlp/extractor/epix.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            'thumbnail': r're:^https?://.+\.jpg',
+        },
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        info = self._download_json(f'{self._API_URL}series/{video_id}', video_id=video_id, headers=self._HEADERS)
+        extras_content = traverse_obj(info, ('series', 'extras', 0, 'content'))
+
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(extras_content.get('hlspath'), video_id=video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:75]: `_extract_m3u8_formats...` will throw error with None. Also, the video cant be downloaded, and no data can be extracted without `extras_content`. So why use traverse_obj/.get here?

Similar issue below

File: yt_dlp/extractor/epix.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'title': extras_content.get('title'),
+            'description': traverse_obj(info, ('series', 'description'), ('series', 'short_description')),
+            'subtitles': subtitles,
+            'thumbnail': traverse_obj(info, ('series', 'images', 'poster', '16_9', 'xlarge')),
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:84]: too specific. It looks like we can extract all thumbnails instead

File: yt_dlp/extractor/epix.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+
+class EpixBaseTraier(EpixBase):
+
+    def _extract_trailer_info(self, info, video_id):
+        if not info.get('trailer'):
+            raise ExtractorError("Video doesn't have trailer", expected=True)
+        formats = [{
+            'url': traverse_obj(info, ('trailer', 'url')),
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:95]: format without url is wrong

File: yt_dlp/extractor/epix.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+            'ext': 'mp4',
+        }]
+        fmts, subtitles = self._extract_m3u8_formats_and_subtitles(traverse_obj(info, ('trailer', 'hlspath')),
+                                                                   video_id=video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:99]: indentation

File: yt_dlp/extractor/epix.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        formats.extend(fmts)
+        self._sort_formats(formats)
+        return {
+            'id': video_id,
+            'title': info.get('title') or info.get('short_title'),
+            'description': info.get('synopsis') or info.get('short_description'),
+            'subtitles': subtitles,
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:106]: traverse_obj/dict_get

File: yt_dlp/extractor/epix.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+            'thumbnail': traverse_obj(info, ('images', 'poster', '16_9', 'xlarge')),
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:107]: same as above

File: yt_dlp/extractor/epix.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+            'description': 'md5:869360873062204eda1aaa2ddc512cdf',
+            'thumbnail': r're:^https?://.+\.jpg',
+        },
+    }]
+
+    def _real_extract(self, url):
+        video_id, season_id, episode_id = self._match_valid_url(url).groups()
+        info = self._download_json(f'{self._API_URL}series/{video_id}', video_id=video_id, headers=self._HEADERS)
+        trailer_info = traverse_obj(info, (
+            'series', 'items', int_or_none(season_id) - 1, 'content', 'items', int_or_none(episode_id) - 1, 'content'))
```
Comment (pukkandan) [inline yt_dlp/extractor/epix.py:142]: why int_or_none if u r going to do `-1`?

**Review by pukkandan — CHANGES_REQUESTED** [review]
Is the actual content DRM protected?
* If yes, the extractor must raise warning mentioning this and that only trailer is being downloaded
* If not, the trailer IEs should not capture non-trailer URLs

File: yt_dlp/extractor/epix.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    int_or_none,
+    traverse_obj,
+)
+
+
+class EpixBase(InfoExtractor):
+    _API_URL = 'https://api.epix.com/v2/'
+    _API_KEY = 'f07debfcdf0f442bab197b517a5126ec'
+    _HEADERS = {
+        'user-agent': 'Mozilla/5.0 (X11; Linux i686) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/65.0.3325.181 Chrome/65.0.3325.181 Safari/537.36',
```
[the remark this answers is not in the document]
Comment (hatienl0i261299) [reply yt_dlp/extractor/epix.py:19]: Yes, it needs.

**Comment by hatienl0i261299** [comment]
Temporarily close PR until no longer geo-blocked

**Pull request closed without merging**